File: activity/20200805/junhaolin/1079.Letter_Tile_Possibilities.cpp
class Solution {
public:
    int numTilePossibilities(string tiles) {
        map<char,int> mp;
        for (int i=0; i<tiles.size(); i++)
            mp[tiles[i]]++;
        
        return count(mp);
    }
    
    int count(map<char,int> mp){
        int sum = 0;
        if (!isEmpty(mp)) {
            for(const auto & [key, value] : mp) { 
                if (value > 0) {
                    mp[key]--;
                    sum ++;
                    sum += count(mp);
                    mp[key]++;
                }
            }
        }    
        return sum;
    }
    
    bool isEmpty(map<char,int> mp) {
        int sum = 0;
        for(const auto & [key, value] : mp)   
            sum += value;
        return sum == 0 ? true : false;
    }
};

File: activity/20200805/ninezero/1079_letter_tile_possibilities.go
func numTilePossibilities(tiles string) int {
	// 统计每一个字母的出现个数
	letters := make([]int, 26)
	for i := 0; i < len(tiles); i++ {
		letters[tiles[i]-'A'] += 1
	}
	return dfs(letters)
}

func dfs(letters []int) int {
	count := 0
	for i, value := range letters {
		if value == 0 {
			continue
		}
		count++
		letters[i]--
		count += dfs(letters)
		letters[i]++
	}
	return count
}
